test/common/network/happy_eyeballs_connection_impl_test.cc

```diff
@@ -0,0 +1,1015 @@
+#include "test/mocks/network/transport_socket.h"
+#include "test/mocks/stream_info/mocks.h"
+
+using testing::Return;
+using testing::ReturnRef;
+using testing::StrictMock;
+
+namespace Envoy {
+namespace Network {
+
+class HappyEyeballsConnectionImplTest : public testing::Test {
+public:
+  HappyEyeballsConnectionImplTest()
+      : failover_timer_(new testing::StrictMock<Event::MockTimer>(&dispatcher_)),
+        transport_socket_options_(std::make_shared<TransportSocketOptionsImpl>()),
+        options_(std::make_shared<ConnectionSocket::Options>()),
+        address_list_({std::make_shared<Address::Ipv4Instance>("127.0.0.1"),
+                       std::make_shared<Address::Ipv4Instance>("127.0.0.2"),
+                       std::make_shared<Address::Ipv4Instance>("127.0.0.3")}) {
+    EXPECT_CALL(transport_socket_factory_, createTransportSocket(_));
+    EXPECT_CALL(dispatcher_, createClientConnection_(address_list_[0], _, _, _))
+        .WillOnce(testing::InvokeWithoutArgs(
+            this, &HappyEyeballsConnectionImplTest::createNextConnection));
+
+    next_connections_.push_back(std::make_unique<StrictMock<MockClientConnection>>());
+    impl_ = std::make_unique<HappyEyeballsConnectionImpl>(
+        dispatcher_, address_list_, Address::InstanceConstSharedPtr(), transport_socket_factory_,
+        transport_socket_options_, options_);
+  }
+
+  // Called by the dispatcher to return a MockClientConnection. In order to allow expectations to
+  // be set on this connection, the object must exist. So instead of allocating a new
+  // MockClientConnection in this method, it instead pops the first entry from
+  // next_connections_ and returns that. It also saves a pointer to that connection into
+  // created_connections_ so that it can be interacted with after it has been returned.
+  MockClientConnection* createNextConnection() {
+    created_connections_.push_back(next_connections_.front().release());
+    next_connections_.pop_front();
+    EXPECT_CALL(*created_connections_.back(), addConnectionCallbacks(_))
+        .WillOnce(
+            Invoke([&](ConnectionCallbacks& cb) -> void { connection_callbacks_.push_back(&cb); }));
+    return created_connections_.back();
+  }
+
+  // Calls connect() on the impl and verifies that the timer is started.
+  void startConnect() {
+    EXPECT_CALL(*failover_timer_, enableTimer(std::chrono::milliseconds(300), nullptr));
+    EXPECT_CALL(*failover_timer_, enabled()).WillRepeatedly(Return(true));
+    EXPECT_CALL(*created_connections_[0], connect());
+    impl_->connect();
+  }
+
+  // Connects the first (and only attempt).
+  void connectFirstAttempt() {
+    startConnect();
+
+    EXPECT_CALL(*failover_timer_, disableTimer());
+    EXPECT_CALL(*created_connections_[0], removeConnectionCallbacks(_));
+    connection_callbacks_[0]->onEvent(ConnectionEvent::Connected);
+  }
+
+  // First the failover timer and creates the next connection.
+  void timeoutAndStartNextAttempt() {
```

Comment:
timeOut?

Reply:
Hm. That still looks weird to me, but since it looks better to you, I'll happily defer to you. Done!